// @modern-js/babel-chain@1.2.1/dist/js/treeshaking/preset.js
function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

export var createBabelPresetChain = function createBabelPresetChain() {
  var presets = [];
  var blacks = [];

  var preset = function preset(name) {
    var isExist = presets.some(function (plugin) {
      return plugin.name === name;
    }); // eslint-disable-next-line @typescript-eslint/no-shadow

    var preset = {
      name: name,
      options: []
    };

    var tap = function tap(options) {
      preset.options = options;
      set();
    };

    var options = function options() {
      return preset.options;
    };

    var del = function del() {
      if (isExist) {
        // eslint-disable-next-line @typescript-eslint/no-shadow
        presets = presets.filter(function (preset) {
          return !preset.name.includes(name);
        });
      }
    };

    var ban = function ban() {
      if (!blacks.includes(name)) {
        blacks.push(name);
      }

      del();
    };

    var set = function set() {
      if (blacks.includes(name)) {
        throw new Error("Preset: ".concat(name, " has been banned!, You can't set it."));
      } // merge preset with replacing
      // see https://babeljs.io/docs/en/configuration#how-babel-merges-config-items


      if (!isExist) {
        presets.push(preset);
      }
    }; // eslint-disable-next-line @typescript-eslint/no-shadow


    var use = function use(path, options) {
      preset.path = path;
      preset.options = options || [];
      set();
    };

    return {
      tap: tap,
      "delete": del,
      ban: ban,
      options: options,
      filter: presets.filter,
      use: use
    };
  };

  var toJSON = function toJSON() {
    return (// eslint-disable-next-line @typescript-eslint/no-shadow
      presets.map(function (preset) {
        return preset.options ? [preset.path || preset.name].concat(_toConsumableArray(preset.options)) : preset.path || preset.name;
      })
    );
  }; // merge preset with replacing
  // see https://babeljs.io/docs/en/configuration#how-babel-merges-config-items


  var merge = function merge(other) {
    // eslint-disable-next-line @typescript-eslint/no-shadow
    var _iterator = _createForOfIteratorHelper(other.presets),
        _step;

    try {
      for (_iterator.s(); !(_step = _iterator.n()).done;) {
        var _preset = _step.value;

        if (_preset.path) {
          chain.preset(_preset.name).use(_preset.path, _preset.options);
        } else {
          chain.preset(_preset.name).tap(_preset.options);
        }
      }
    } catch (err) {
      _iterator.e(err);
    } finally {
      _iterator.f();
    }

    return chain;
  };

  var chain = {
    preset: preset,
    presets: presets,
    blacks: blacks,
    toJSON: toJSON,
    merge: merge
  };
  return chain;
};
