// @modern-js/runtime-core@1.2.4/dist/js/modern/plugin.js
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import { createManager, createPipeline, createAsyncPipeline, createContext } from '@modern-js/plugin';
import { createLoaderManager } from "./loader/loaderManager";
const hoc = createPipeline(); // eslint-disable-next-line @typescript-eslint/no-empty-interface

const provide = createPipeline();
export const AppComponentContext = createContext(null);
export const useAppComponent = () => {
  const AppComponent = AppComponentContext.use().value;

  if (!AppComponent) {
    throw new Error(`Expect React.ComponentType, accept: null`);
  }

  return AppComponent;
};
export const useRootElement = () => {
  const rootElement = AppComponentContext.use().value;

  if (!rootElement) {
    throw new Error(`Expect HTMLElement, accept: null`);
  }

  return rootElement;
};
const client = createAsyncPipeline();
const server = createAsyncPipeline();
const init = createAsyncPipeline();
const pickContext = createPipeline();
export const createRuntime = () => createManager({
  hoc,
  provide,
  client,
  server,
  init,
  pickContext
});
/**
 * register init hook. It would be revoked both ssr and csr.
 */

const registerInit = (App, _init) => {
  const originalInit = App.init;

  App.init = async context => {
    if (!context.loaderManager) {
      context.loaderManager = createLoaderManager({});
    }

    await Promise.all([originalInit === null || originalInit === void 0 ? void 0 : originalInit(context), _init === null || _init === void 0 ? void 0 : _init(context)]);
  };
};
/**
 * register prefetch hook. It would be revoked both ssr and csr.
 * But if ssr success, It wont exec in csr again.
 * If ssr prefetch failed, It will fallback to exec in csr.
 */


const registerPrefetch = (App, prefetch) => {
  const originalPrefetch = App.prefetch;

  App.prefetch = async context => {
    const originResult = await (originalPrefetch === null || originalPrefetch === void 0 ? void 0 : originalPrefetch(context));
    const result = await (prefetch === null || prefetch === void 0 ? void 0 : prefetch(context));
    return _objectSpread(_objectSpread({}, originResult), result);
  };
};

export const runtime = createRuntime();
export const {
  createPlugin
} = runtime;
export { registerInit, registerPrefetch };
