// @modern-js/css-config@1.2.2/dist/js/modern/postcss.js
import { applyOptionsChain, getBrowserslist } from '@modern-js/utils';
import { shouldUseSourceMap } from "./util";
/**
 *
 * @param autoprefixer - Enable autoprefixer config.
 * @returns Base PostCSS config options for app or library.
 */

export const getPostcssConfig = (appDirectory, config, autoprefixer = true) => {
  var _config$tools, _config$tools2;

  return applyOptionsChain({
    postcssOptions: {
      plugins: [require('postcss-flexbugs-fixes'), require('postcss-custom-properties'), require('postcss-initial'), require('postcss-page-break'), require('postcss-font-variant'), require('postcss-media-minmax'), require('postcss-nesting'), autoprefixer ? require('autoprefixer')(applyOptionsChain({
        flexbox: 'no-2009',
        overrideBrowserslist: getBrowserslist(appDirectory)
      }, (_config$tools = config.tools) === null || _config$tools === void 0 ? void 0 : _config$tools.autoprefixer)) : false].filter(Boolean)
    },
    sourceMap: shouldUseSourceMap(config)
  }, (_config$tools2 = config.tools) === null || _config$tools2 === void 0 ? void 0 : _config$tools2.postcss);
};
