// @modern-js/core@1.4.6/dist/js/node/config/schema/source.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.source = void 0;

var _utils = require("@modern-js/utils");

const source = {
  type: 'object',
  additionalProperties: false,
  properties: {
    entries: {
      type: 'object',
      patternProperties: {
        [_utils.ENTRY_NAME_PATTERN]: {
          if: {
            type: 'object'
          },
          then: {
            required: ['entry'],
            properties: {
              entry: {
                type: ['string', 'array']
              },
              disableMount: {
                type: 'boolean'
              },
              enableFileSystemRoutes: {
                type: 'boolean'
              }
            },
            additionalProperties: false
          },
          else: {
            type: ['string', 'array']
          }
        }
      }
    },
    alias: {
      typeof: ['object', 'function']
    },
    disableDefaultEntries: {
      type: 'boolean'
    },
    envVars: {
      type: 'array'
    },
    globalVars: {
      type: 'object'
    },
    moduleScopes: {
      instanceof: ['Array', 'Function']
    },
    entriesDir: {
      type: 'string'
    },
    configDir: {
      type: 'string'
    },
    include: {
      type: ['array']
    }
  }
};
exports.source = source;
