// @modern-js/runtime@1.2.3/dist/js/node/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "RuntimeReactContext", {
  enumerable: true,
  get: function () {
    return _runtimeCore.RuntimeReactContext;
  }
});
Object.defineProperty(exports, "bootstrap", {
  enumerable: true,
  get: function () {
    return _runtimeCore.bootstrap;
  }
});
Object.defineProperty(exports, "createApp", {
  enumerable: true,
  get: function () {
    return _runtimeCore.createApp;
  }
});
Object.defineProperty(exports, "createPlugin", {
  enumerable: true,
  get: function () {
    return _runtimeCore.createPlugin;
  }
});
Object.defineProperty(exports, "defineConfig", {
  enumerable: true,
  get: function () {
    return _runtimeCore.defineConfig;
  }
});
Object.defineProperty(exports, "registerInit", {
  enumerable: true,
  get: function () {
    return _runtimeCore.registerInit;
  }
});
Object.defineProperty(exports, "registerPrefetch", {
  enumerable: true,
  get: function () {
    return _runtimeCore.registerPrefetch;
  }
});
Object.defineProperty(exports, "useLoader", {
  enumerable: true,
  get: function () {
    return _runtimeCore.useLoader;
  }
});
Object.defineProperty(exports, "useRuntimeContext", {
  enumerable: true,
  get: function () {
    return _runtimeCore.useRuntimeContext;
  }
});

var _runtimeCore = require("@modern-js/runtime-core");
