// @modern-js/plugin-state@1.2.1/dist/js/node/runtime/plugin.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {};
exports.default = void 0;

var _react = require("react");

var _runtimeCore = require("@modern-js/runtime-core");

var _store = require("@modern-js-reduck/store");

var _react2 = require("@modern-js-reduck/react");

var _hoistNonReactStatics = _interopRequireDefault(require("hoist-non-react-statics"));

var _jsxRuntime = require("react/jsx-runtime");

var _plugins = require("../plugins");

Object.keys(_plugins).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _plugins[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _plugins[key];
    }
  });
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const state = config => (0, _runtimeCore.createPlugin)(() => ({
  hoc({
    App
  }, next) {
    const getStateApp = props => {
      // eslint-disable-next-line react-hooks/rules-of-hooks
      const context = (0, _react.useContext)(_runtimeCore.RuntimeReactContext);
      return /*#__PURE__*/(0, _jsxRuntime.jsx)(_react2.Provider, {
        store: context.store,
        config: config,
        children: /*#__PURE__*/(0, _jsxRuntime.jsx)(App, _objectSpread({}, props))
      });
    };

    return next({
      App: (0, _hoistNonReactStatics.default)(getStateApp, App)
    });
  },

  init({
    context
  }, next) {
    const storeConfig = config || {};

    if (typeof window !== 'undefined') {
      var _window, _window$_SSR_DATA, _window$_SSR_DATA$dat;

      storeConfig.initialState = storeConfig.initialState || ((_window = window) === null || _window === void 0 ? void 0 : (_window$_SSR_DATA = _window._SSR_DATA) === null || _window$_SSR_DATA === void 0 ? void 0 : (_window$_SSR_DATA$dat = _window$_SSR_DATA.data) === null || _window$_SSR_DATA$dat === void 0 ? void 0 : _window$_SSR_DATA$dat.storeState) || {};
    }

    context.store = (0, _store.createStore)(storeConfig);
    next({
      context
    });
  },

  pickContext({
    context,
    pickedContext
  }, next) {
    return next({
      context,
      pickedContext: _objectSpread(_objectSpread({}, pickedContext), {}, {
        store: context.store
      })
    });
  }

}), {
  name: '@modern-js/plugin-state'
});

var _default = state;
exports.default = _default;
