// @modern-js/plugin-fast-refresh@1.2.1/dist/js/node/cli.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _core = require("@modern-js/core");

var _utils = require("@modern-js/utils");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var _default = (0, _core.createPlugin)(() => ({
  config() {
    return {
      tools: {
        webpack: (config, {
          chain,
          name
        }) => {
          if (name === 'client' && (0, _utils.isFastRefresh)()) {
            const ReactFastRefreshPlugin = require('@pmmmwh/react-refresh-webpack-plugin');

            chain.plugin('react-fast-refresh').use(ReactFastRefreshPlugin, [{
              overlay: false,
              exclude: [/node_modules/]
            }]);
            const loaders = chain.module.rule('loaders');
            const babelOptions = loaders.oneOf('js').use('babel').get('options');
            loaders.oneOf('js').use('babel').options(_objectSpread(_objectSpread({}, babelOptions), {}, {
              plugins: [...(babelOptions.plugins || []), require.resolve('react-refresh/babel')]
            }));
          }
        }
      }
    };
  }

}), {
  name: '@modern-js/plugin-fast-refresh'
});

exports.default = _default;
