// @modern-js/codesmith@1.0.8/dist/js/node/materials/FsMaterial.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FsMaterial = void 0;

var _path = _interopRequireDefault(require("path"));

var _globPromise = _interopRequireDefault(require("glob-promise"));

var _FsResource = require("./FsResource");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class FsMaterial {
  constructor(basePath) {
    this.basePath = void 0;
    this.basePath = basePath;
  }

  get(resourceKey) {
    return new _FsResource.FsResource(_path.default.resolve(this.basePath, resourceKey), resourceKey);
  }

  async find(globStr, options) {
    const matches = await (0, _globPromise.default)(globStr, {
      cwd: _path.default.resolve(this.basePath),
      nodir: options === null || options === void 0 ? void 0 : options.nodir,
      dot: options === null || options === void 0 ? void 0 : options.dot,
      ignore: options === null || options === void 0 ? void 0 : options.ignore
    });
    return matches.reduce((pre, cur) => {
      pre[cur] = new _FsResource.FsResource(_path.default.resolve(this.basePath, cur), cur);
      return pre;
    }, {});
  }

}

exports.FsMaterial = FsMaterial;
