// @modern-js/codesmith-api-git@1.0.7/dist/js/node/utils/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.canUseGit = canUseGit;
exports.gitAdd = gitAdd;
exports.gitCommit = gitCommit;
exports.initGitRepo = initGitRepo;
exports.isInGitRepo = isInGitRepo;

var _execa = _interopRequireDefault(require("execa"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

async function canUseGit() {
  try {
    await (0, _execa.default)('git', ['--version'], {
      env: process.env
    });
    return true;
  } catch (e) {
    return false;
  }
}

async function isInGitRepo(cwd) {
  try {
    await (0, _execa.default)('git', ['rev-parse', '--is-inside-work-tree'], {
      env: process.env,
      cwd
    });
    return true;
  } catch (e) {
    return false;
  }
}

async function initGitRepo(cwd, defaultBranch) {
  await (0, _execa.default)('git', ['init'], {
    env: process.env,
    cwd
  });
  const {
    stdout
  } = await (0, _execa.default)('git', ['symbolic-ref', '--short', 'HEAD'], {
    env: process.env,
    cwd
  });

  if (stdout !== defaultBranch) {
    await (0, _execa.default)('git', ['checkout', '-b', defaultBranch], {
      env: process.env,
      cwd
    });
  }
}

async function gitAdd(cwd) {
  await (0, _execa.default)('git', ['add', '-A'], {
    env: process.env,
    cwd
  });
}

async function gitCommit(cwd, commitMessage) {
  await (0, _execa.default)('git', ['commit', '-m', commitMessage, '--no-verify'], {
    env: process.env,
    cwd
  });
}
