// @modern-js/utils@1.3.4/dist/js/node/alias.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.validAlias = exports.getUserAlias = exports.getAlias = void 0;

var _fs = _interopRequireDefault(require("fs"));

var _path = _interopRequireDefault(require("path"));

var _chalk = _interopRequireDefault(require("chalk"));

var _readTsConfig = require("./readTsConfig");

var _applyOptionsChain = require("./applyOptionsChain");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const validAlias = (modernConfig, {
  tsconfigPath
}) => {
  const {
    source: {
      alias
    }
  } = modernConfig;

  if (!alias) {
    return null;
  }

  const isTsProject = _fs.default.existsSync(tsconfigPath);

  if (!isTsProject) {
    return null;
  }

  const userAlias = getUserAlias(alias);

  if (Object.keys(userAlias).length > 0) {
    return _chalk.default.red('Note: Please use `compilerOptions.paths` in "tsconfig.json" file replace `source.alias` config in "modern.config.js/ts" when project is typescript');
  }

  return null;
};

exports.validAlias = validAlias;

const getAlias = (aliasOption, option) => {
  const isTsProject = _fs.default.existsSync(option.tsconfigPath);

  let aliasConfig;

  if (!isTsProject) {
    aliasConfig = {
      absoluteBaseUrl: option.appDirectory,
      paths: (0, _applyOptionsChain.applyOptionsChain)({
        '@': ['./src']
      }, aliasOption),
      isTsPath: false,
      isTsProject
    };
  } else {
    var _tsconfig$compilerOpt, _tsconfig$compilerOpt2;

    const tsconfig = (0, _readTsConfig.readTsConfigByFile)(option.tsconfigPath);
    const baseUrl = tsconfig === null || tsconfig === void 0 ? void 0 : (_tsconfig$compilerOpt = tsconfig.compilerOptions) === null || _tsconfig$compilerOpt === void 0 ? void 0 : _tsconfig$compilerOpt.baseUrl;
    aliasConfig = {
      absoluteBaseUrl: baseUrl ? _path.default.join(option.appDirectory, baseUrl) : option.appDirectory,
      paths: _objectSpread(_objectSpread({}, aliasOption || {}), tsconfig === null || tsconfig === void 0 ? void 0 : (_tsconfig$compilerOpt2 = tsconfig.compilerOptions) === null || _tsconfig$compilerOpt2 === void 0 ? void 0 : _tsconfig$compilerOpt2.paths),
      isTsPath: true,
      isTsProject
    };
  }

  return aliasConfig;
}; // filter invalid ts paths that are not array


exports.getAlias = getAlias;

const getUserAlias = (alias = {}) => Object.keys(alias).reduce((o, k) => {
  if (Array.isArray(alias[k])) {
    o[k] = alias[k];
  }

  return o;
}, {});

exports.getUserAlias = getUserAlias;
