// @modern-js/utils@1.3.4/dist/js/node/watch.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.watch = exports.WatchChangeType = void 0;

var path = _interopRequireWildcard(require("path"));

var _import = require("./import");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const chokidar = _import.Import.lazy('chokidar', require);

const WatchChangeType = {
  ADD: 'add',
  UNLINK: 'unlink',
  CHANGE: 'change'
};
exports.WatchChangeType = WatchChangeType;

const watch = (watchDir, runTask, ignored = []) => {
  let ready = false;
  const watcher = chokidar.watch(watchDir, {
    ignored
  });
  watcher.on('ready', () => ready = true);
  watcher.on('change', async filePath => {
    const changedFilePath = path.resolve(filePath);
    await runTask({
      changedFilePath,
      changeType: WatchChangeType.CHANGE
    });
  });
  watcher.on('add', async filePath => {
    const changedFilePath = path.resolve(filePath);

    if (ready) {
      await runTask({
        changedFilePath,
        changeType: WatchChangeType.ADD
      });
    }
  });
  watcher.on('unlink', async filePath => {
    const changedFilePath = path.resolve(filePath);
    await runTask({
      changedFilePath,
      changeType: WatchChangeType.UNLINK
    });
  });
  watcher.on('error', err => {
    throw err;
  });
  return watcher;
};

exports.watch = watch;
