// @modern-js/plugin-ssr@1.2.3/dist/js/modern/index.node.js
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

// eslint-disable-next-line filenames/match-exported
import path from 'path';
import { createPlugin, registerPrefetch } from '@modern-js/runtime-core';
import prefetch from "./prefetch";
export { run, useHeaders } from "./hook";
const registeredApps = new WeakSet();

const plugin = () => createPlugin(() => ({
  server: async ({
    App,
    context
  }) => {
    if (!registeredApps.has(App)) {
      registerPrefetch(App, _context => prefetch(App, _context));
      registeredApps.add(App);
    }

    if (typeof window === 'undefined') {
      const html = await require("./serverRender").render(context, (context === null || context === void 0 ? void 0 : context.ssrContext.distDir) || path.join(process.cwd(), 'dist'), App);
      return html;
    }

    return null;
  },
  pickContext: ({
    context,
    pickedContext
  }, next) => {
    const {
      request
    } = context === null || context === void 0 ? void 0 : context.ssrContext;
    const {
      cookie,
      'user-agent': userAgent,
      referer
    } = request.headers || {};
    return next({
      context,
      pickedContext: _objectSpread(_objectSpread({}, pickedContext), {}, {
        request: _objectSpread({
          cookie,
          userAgent,
          referer
        }, request) // FIXME: error TS2322: Type '{ request: any; store: Store<any, AnyAction> & { use: UseModel; }; }' is not assignable to type 'TRuntimeContext'. Object literal may only specify known properties, and 'request' does not exist in type 'TRuntimeContext'.

      })
    });
  }
}), {
  name: '@modern-js/plugin-ssr'
});

export default plugin;
export * from "./react";
