// @modern-js/app-tools@1.4.3/dist/js/node/utils/getSpecifiedEntries.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getSpecifiedEntries = void 0;

var _inquirer = _interopRequireDefault(require("inquirer"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const getSpecifiedEntries = async (entry, entrypoints) => {
  const entryNames = entrypoints.map(e => e.entryName);

  if (!entry) {
    return entryNames;
  }

  if (typeof entry === 'boolean') {
    const {
      selected
    } = await _inquirer.default.prompt([{
      type: 'checkbox',
      name: 'selected',
      choices: entryNames,
      message: '请选择需要构建的入口',

      validate(answer) {
        if (answer.length < 1) {
          return 'You must choose at least one topping.';
        }

        return true;
      }

    }]);
    return selected;
  }

  entry.forEach(name => {
    if (!entryNames.includes(name)) {
      throw new Error(`can not found entry ${name}, compiler entry should in ${entryNames.join(', ')}`);
    }
  });
  return entry;
};

exports.getSpecifiedEntries = getSpecifiedEntries;
