// @modern-js/prod-server@1.0.3/dist/js/modern/libs/hook-api/template.js
const RegList = {
  before: {
    head: '<head>',
    body: '<body>'
  },
  after: {
    head: '</head>',
    body: '</body>'
  }
};

class TemplateAPI {
  constructor(content) {
    this.content = void 0;
    this.content = content;
  }

  get() {
    return this.content;
  }

  set(content) {
    this.content = content;
  }

  prependHead(fragment) {
    const {
      head
    } = RegList.before;
    return this.replace(head, `${head}${fragment}`);
  }

  appendHead(fragment) {
    const {
      head
    } = RegList.after;
    return this.replace(head, `${fragment}${head}`);
  }

  prependBody(fragment) {
    const {
      body
    } = RegList.before;
    return this.replace(body, `${body}${fragment}`);
  }

  appendBody(fragment) {
    const {
      body
    } = RegList.after;
    return this.replace(body, `${fragment}${body}`);
  }

  replace(reg, text) {
    this.content = this.content.replace(reg, text);
    return this;
  }

}

export const createTemplateAPI = content => new TemplateAPI(content);
