// @modern-js/utils@1.3.4/dist/js/treeshaking/constants.js
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * hmr socket connect path
 */
export var HMR_SOCK_PATH = '/_modern_js_hmr_ws';
/**
 * route specification file
 */

export var ROUTE_SPEC_FILE = 'route.json';
/**
 * main entry name
 */

export var MAIN_ENTRY_NAME = 'main';
/**
 * open editor request path
 */

export var LAUNCH_EDITOR_ENDPOINT = '/__open-stack-frame-in-editor';
/**
 * server side bundles directory, which relative to dist.
 */

export var SERVER_BUNDLE_DIRECTORY = 'bundles';
/**
 * entry name pattern used for ajv pattern properties.
 */

export var ENTRY_NAME_PATTERN = '^[a-zA-Z0-9_-]+$';
/**
 * SSR server render function name
 */

export var SERVER_RENDER_FUNCTION_NAME = 'serverRender';
/**
 * loadbale manifest json file
 */

export var LOADABLE_STATS_FILE = 'loadable-stats.json';
/**
 * internal specified folder
 */

export var API_DIR = 'api';
export var SERVER_DIR = 'server';
export var SHARED_DIR = 'shared';
/**
 * Internal plugins that work as soon as they are installed.
 */

export var INTERNAL_PLUGINS = {
  '@modern-js/app-tools': {
    cli: '@modern-js/app-tools/cli'
  },
  '@modern-js/monorepo-tools': {
    cli: '@modern-js/monorepo-tools/cli'
  },
  '@modern-js/module-tools': {
    cli: '@modern-js/module-tools/cli'
  },
  '@modern-js/runtime': {
    cli: '@modern-js/runtime/cli'
  },
  '@modern-js/plugin-less': {
    cli: '@modern-js/plugin-less/cli'
  },
  '@modern-js/plugin-sass': {
    cli: '@modern-js/plugin-sass/cli'
  },
  '@modern-js/plugin-esbuild': {
    cli: '@modern-js/plugin-esbuild/cli'
  },
  '@modern-js/plugin-proxy': {
    cli: '@modern-js/plugin-proxy/cli'
  },
  '@modern-js/plugin-ssg': {
    cli: '@modern-js/plugin-ssg/cli'
  },
  '@modern-js/plugin-bff': {
    cli: '@modern-js/plugin-bff/cli',
    server: '@modern-js/plugin-bff/server'
  },
  '@modern-js/plugin-electron': {
    cli: '@modern-js/plugin-electron/cli'
  },
  '@modern-js/plugin-testing': {
    cli: '@modern-js/plugin-testing/cli'
  },
  '@modern-js/plugin-storybook': {
    cli: '@modern-js/plugin-storybook/cli'
  },
  '@modern-js/plugin-docsite': {
    cli: '@modern-js/plugin-docsite/cli'
  },
  '@modern-js/plugin-express': {
    cli: '@modern-js/plugin-express/cli',
    server: '@modern-js/plugin-express'
  },
  '@modern-js/plugin-egg': {
    cli: '@modern-js/plugin-egg/cli',
    server: '@modern-js/plugin-egg'
  },
  '@modern-js/plugin-koa': {
    cli: '@modern-js/plugin-koa/cli',
    server: '@modern-js/plugin-koa'
  },
  '@modern-js/plugin-nest': {
    cli: '@modern-js/plugin-nest/cli',
    server: '@modern-js/plugin-nest/server'
  },
  '@modern-js/plugin-unbundle': {
    cli: '@modern-js/plugin-unbundle'
  },
  '@modern-js/plugin-server-build': {
    cli: '@modern-js/plugin-server-build'
  },
  '@modern-js/plugin-server': {
    cli: '@modern-js/plugin-server/cli',
    server: '@modern-js/plugin-server/server'
  },
  '@modern-js/plugin-garfish': {
    cli: '@modern-js/plugin-garfish/cli'
  },
  '@modern-js/plugin-jarvis': {
    cli: '@modern-js/plugin-jarvis/cli'
  },
  '@modern-js/plugin-tailwindcss': {
    cli: '@modern-js/plugin-tailwindcss/cli'
  },
  '@modern-js/plugin-lambda-fc': {
    cli: '@modern-js/plugin-lambda-fc/cli'
  },
  '@modern-js/plugin-lambda-scf': {
    cli: '@modern-js/plugin-lambda-scf/cli'
  },
  '@modern-js/plugin-cdn-oss': {
    cli: '@modern-js/plugin-cdn-oss/cli'
  },
  '@modern-js/plugin-cdn-cos': {
    cli: '@modern-js/plugin-cdn-cos/cli'
  },
  '@modern-js/plugin-static-hosting': {
    cli: '@modern-js/plugin-static-hosting/cli'
  },
  '@modern-js/plugin-polyfill': {
    cli: '@modern-js/plugin-polyfill/cli',
    server: '@modern-js/plugin-polyfill'
  },
  '@modern-js/plugin-multiprocess': {
    cli: '@modern-js/plugin-multiprocess/cli'
  },
  '@modern-js/plugin-nocode': {
    cli: '@modern-js/plugin-nocode/cli'
  },
  '@modern-js/plugin-design-token': {
    cli: '@modern-js/plugin-design-token/cli'
  }
};
/**
 * The schema registered in the plugin.
 */

export var PLUGIN_SCHEMAS = {
  '@modern-js/runtime': [{
    target: 'runtime',
    schema: {
      type: 'object',
      additionalProperties: false
    }
  }, {
    target: 'runtimeByEntries',
    schema: {
      type: 'object',
      patternProperties: _defineProperty({}, ENTRY_NAME_PATTERN, {
        type: 'object'
      }),
      additionalProperties: false
    }
  }],
  '@modern-js/plugin-bff': [{
    target: 'bff',
    schema: {
      type: 'object',
      properties: {
        prefix: {
          type: ['string', 'array'],
          items: {
            type: 'string'
          }
        },
        fetcher: {
          type: 'string'
        },
        proxy: {
          type: 'object'
        },
        requestCreator: {
          type: 'string'
        }
      }
    }
  }],
  '@modern-js/plugin-esbuild': [{
    target: 'tools.esbuild',
    schema: {
      "typeof": ['object']
    }
  }],
  '@modern-js/plugin-less': [{
    target: 'tools.less',
    schema: {
      "typeof": ['object', 'function']
    }
  }],
  '@modern-js/plugin-sass': [{
    target: 'tools.sass',
    schema: {
      "typeof": ['object', 'function']
    }
  }],
  '@modern-js/plugin-tailwindcss': [{
    target: 'tools.tailwindcss',
    schema: {
      "typeof": ['object', 'function']
    }
  }],
  '@modern-js/plugin-proxy': [{
    target: 'dev.proxy',
    schema: {
      "typeof": ['string', 'object']
    }
  }],
  '@modern-js/plugin-unbundle': [{
    target: 'source.disableAutoImportStyle',
    schema: {
      type: 'boolean'
    }
  }, {
    target: 'server.https',
    schema: {
      type: 'boolean'
    }
  }, {
    target: 'dev.unbundle',
    schema: {
      type: 'object',
      properties: {
        ignore: {
          type: ['string', 'array'],
          items: {
            type: 'string'
          }
        },
        ignoreModuleCache: {
          type: 'boolean'
        },
        clearPdnCache: {
          type: 'boolean'
        },
        pdnHost: {
          type: 'string'
        }
      }
    }
  }],
  '@modern-js/plugin-ssg': [{
    target: 'output.ssg',
    schema: {
      oneOf: [{
        type: 'boolean'
      }, {
        type: 'object'
      }, {
        "instanceof": 'Function'
      }]
    }
  }],
  '@modern-js/plugin-ssr': [{
    target: 'runtime.ssr',
    schema: {
      type: ['boolean', 'object']
    }
  }],
  '@modern-js/plugin-state': [{
    target: 'runtime.state',
    schema: {
      type: ['boolean', 'object']
    }
  }],
  '@modern-js/plugin-design-token': [// Legacy Features
  {
    target: 'source.designSystem',
    schema: {
      "typeof": ['object']
    }
  }, {
    target: 'source.designSystem.supportStyledComponents',
    schema: {
      type: ['boolean']
    }
  }],
  '@modern-js/plugin-router': [{
    target: 'runtime.router',
    schema: {
      type: ['boolean', 'object']
    }
  }],
  '@modern-js/plugin-testing': [{
    target: 'testing',
    schema: {
      "typeof": ['object']
    }
  }, {
    target: 'tools.jest',
    schema: {
      "typeof": ['object', 'function']
    }
  }],
  '@modern-js/plugin-garfish': [{
    target: 'runtime.masterApp',
    schema: {
      type: ['boolean', 'object']
    }
  }, {
    target: 'dev.withMasterApp',
    schema: {
      type: ['object']
    }
  }, {
    target: 'deploy.microFrontend',
    schema: {
      type: ['boolean', 'object']
    }
  }],
  '@modern-js/plugin-nocode': []
};
