// @modern-js/codesmith@1.0.8/dist/js/modern/logger/index.js
// eslint-disable-next-line eslint-comments/disable-enable-pair

/* eslint-disable no-console */
import chalk from 'chalk';
import { LevelPriority, LoggerLevel } from "./constants";
export class Logger {
  constructor(level = LoggerLevel.Info) {
    this.level = LoggerLevel.Info;
    this.level = level;
  }

  get currentLevelIndex() {
    return LevelPriority.indexOf(this.level);
  }

  getLevalIndex(level) {
    return LevelPriority.indexOf(level);
  }

  error(...meta) {
    if (this.currentLevelIndex < this.getLevalIndex(LoggerLevel.Error)) {
      return;
    }

    console.log(chalk.red('[ERROR]'), ...meta);
  }

  warn(...meta) {
    if (this.currentLevelIndex < this.getLevalIndex(LoggerLevel.Warn)) {
      return;
    }

    console.log(chalk.yellow('[WARN]'), ...meta);
  }

  info(...meta) {
    if (this.currentLevelIndex < this.getLevalIndex(LoggerLevel.Info)) {
      return;
    }

    console.log(chalk.green('[INFO]'), ...meta);
  }

  debug(...meta) {
    if (this.currentLevelIndex < this.getLevalIndex(LoggerLevel.Debug)) {
      return;
    }

    console.log(chalk.blue('[DEBUG]'), ...meta);
  }

  verbose(...meta) {
    if (this.currentLevelIndex < this.getLevalIndex(LoggerLevel.Verbose)) {
      return;
    }

    console.log('[VERBOSE]', ...meta);
  }

  stream(...meta) {
    if (this.currentLevelIndex < this.getLevalIndex(LoggerLevel.Stream)) {
      return;
    }

    console.log('[STREAM]', ...meta);
  }

}
