// @modern-js/app-tools@1.4.3/dist/js/node/lifecycle.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.lifecycle = exports.beforePrintInstructions = exports.beforeDev = exports.beforeDeploy = exports.beforeCreateCompiler = exports.beforeBuild = exports.afterDev = exports.afterDeploy = exports.afterCreateCompiler = exports.afterBuild = void 0;

var _plugin = require("@modern-js/plugin");

var _core = require("@modern-js/core");

const beforeDev = (0, _plugin.createAsyncWorkflow)();
exports.beforeDev = beforeDev;
const afterDev = (0, _plugin.createAsyncWorkflow)();
exports.afterDev = afterDev;
const beforeCreateCompiler = (0, _plugin.createAsyncWorkflow)();
exports.beforeCreateCompiler = beforeCreateCompiler;
const afterCreateCompiler = (0, _plugin.createAsyncWorkflow)();
exports.afterCreateCompiler = afterCreateCompiler;
const beforePrintInstructions = (0, _plugin.createAsyncWaterfall)();
exports.beforePrintInstructions = beforePrintInstructions;
const beforeBuild = (0, _plugin.createAsyncWorkflow)();
exports.beforeBuild = beforeBuild;
const afterBuild = (0, _plugin.createAsyncWorkflow)();
exports.afterBuild = afterBuild;
const beforeDeploy = (0, _plugin.createAsyncWorkflow)();
exports.beforeDeploy = beforeDeploy;
const afterDeploy = (0, _plugin.createAsyncWorkflow)();
exports.afterDeploy = afterDeploy;

const lifecycle = () => {
  (0, _core.registerHook)({
    beforeDev,
    afterDev,
    beforeCreateCompiler,
    afterCreateCompiler,
    beforePrintInstructions,
    beforeBuild,
    afterBuild,
    beforeDeploy,
    afterDeploy
  });
};

exports.lifecycle = lifecycle;
