// @modern-js/css-config@1.2.2/dist/js/node/postcss.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getPostcssConfig = void 0;

var _utils = require("@modern-js/utils");

var _util = require("./util");

/**
 *
 * @param autoprefixer - Enable autoprefixer config.
 * @returns Base PostCSS config options for app or library.
 */
const getPostcssConfig = (appDirectory, config, autoprefixer = true) => {
  var _config$tools, _config$tools2;

  return (0, _utils.applyOptionsChain)({
    postcssOptions: {
      plugins: [require('postcss-flexbugs-fixes'), require('postcss-custom-properties'), require('postcss-initial'), require('postcss-page-break'), require('postcss-font-variant'), require('postcss-media-minmax'), require('postcss-nesting'), autoprefixer ? require('autoprefixer')((0, _utils.applyOptionsChain)({
        flexbox: 'no-2009',
        overrideBrowserslist: (0, _utils.getBrowserslist)(appDirectory)
      }, (_config$tools = config.tools) === null || _config$tools === void 0 ? void 0 : _config$tools.autoprefixer)) : false].filter(Boolean)
    },
    sourceMap: (0, _util.shouldUseSourceMap)(config)
  }, (_config$tools2 = config.tools) === null || _config$tools2 === void 0 ? void 0 : _config$tools2.postcss);
};

exports.getPostcssConfig = getPostcssConfig;
