// @modern-js/runtime-core@1.2.4/dist/js/node/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  useAppComponent: true,
  useRootElement: true,
  createPlugin: true,
  createRuntime: true,
  runtime: true,
  registerInit: true,
  registerPrefetch: true,
  defineConfig: true,
  getConfig: true,
  RuntimeReactContext: true
};
Object.defineProperty(exports, "RuntimeReactContext", {
  enumerable: true,
  get: function () {
    return _runtimeContext.RuntimeReactContext;
  }
});
Object.defineProperty(exports, "createPlugin", {
  enumerable: true,
  get: function () {
    return _plugin.createPlugin;
  }
});
Object.defineProperty(exports, "createRuntime", {
  enumerable: true,
  get: function () {
    return _plugin.createRuntime;
  }
});
Object.defineProperty(exports, "defineConfig", {
  enumerable: true,
  get: function () {
    return _appConfig.defineConfig;
  }
});
Object.defineProperty(exports, "getConfig", {
  enumerable: true,
  get: function () {
    return _appConfig.getConfig;
  }
});
Object.defineProperty(exports, "registerInit", {
  enumerable: true,
  get: function () {
    return _plugin.registerInit;
  }
});
Object.defineProperty(exports, "registerPrefetch", {
  enumerable: true,
  get: function () {
    return _plugin.registerPrefetch;
  }
});
Object.defineProperty(exports, "runtime", {
  enumerable: true,
  get: function () {
    return _plugin.runtime;
  }
});
Object.defineProperty(exports, "useAppComponent", {
  enumerable: true,
  get: function () {
    return _plugin.useAppComponent;
  }
});
Object.defineProperty(exports, "useRootElement", {
  enumerable: true,
  get: function () {
    return _plugin.useRootElement;
  }
});

var _wrap = require("./wrap");

Object.keys(_wrap).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _wrap[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _wrap[key];
    }
  });
});

var _render = require("./render");

Object.keys(_render).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _render[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _render[key];
    }
  });
});

var _initial = require("./initial");

Object.keys(_initial).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _initial[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _initial[key];
    }
  });
});

var _plugin = require("./plugin");

var _appConfig = require("./app-config");

var _compatible = require("./compatible");

Object.keys(_compatible).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _compatible[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _compatible[key];
    }
  });
});

var _runtimeContext = require("./runtime-context");

var _loader = require("./loader");

Object.keys(_loader).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _loader[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _loader[key];
    }
  });
});

var _plugin2 = require("@modern-js/plugin");

Object.keys(_plugin2).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _plugin2[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _plugin2[key];
    }
  });
});
