// @modern-js/bff-utils@1.2.2/dist/js/node/constant.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.INTROSPECTION_ROUTE_PATH = exports.INTROSPECTION_ROUTE_METHOD = exports.INDEX_SUFFIX = exports.HttpMethod = exports.FRAMEWORK_MODE_LAMBDA_DIR = exports.AllHttpMethods = exports.API_DIR = exports.APIMode = void 0;
let HttpMethod;
exports.HttpMethod = HttpMethod;

(function (HttpMethod) {
  HttpMethod["GET"] = "GET";
  HttpMethod["POST"] = "POST";
  HttpMethod["PUT"] = "PUT";
  HttpMethod["DELETE"] = "DELETE";
  HttpMethod["CONNECT"] = "CONNECT";
  HttpMethod["TRACE"] = "TRACE";
  HttpMethod["PATCH"] = "PATCH";
  HttpMethod["OPTION"] = "OPTION";
  HttpMethod["HEAD"] = "HEAD";
})(HttpMethod || (exports.HttpMethod = HttpMethod = {}));

const AllHttpMethods = Object.values(HttpMethod);
exports.AllHttpMethods = AllHttpMethods;
let APIMode;
exports.APIMode = APIMode;

(function (APIMode) {
  APIMode["FARMEWORK"] = "FARMEWORK";
  APIMode["FUNCTION"] = "FUNCTION";
})(APIMode || (exports.APIMode = APIMode = {}));

const FRAMEWORK_MODE_LAMBDA_DIR = 'lambda';
exports.FRAMEWORK_MODE_LAMBDA_DIR = FRAMEWORK_MODE_LAMBDA_DIR;
const INDEX_SUFFIX = 'index';
exports.INDEX_SUFFIX = INDEX_SUFFIX;
const INTROSPECTION_ROUTE_PATH = '/__introspection__';
exports.INTROSPECTION_ROUTE_PATH = INTROSPECTION_ROUTE_PATH;
const API_DIR = 'api';
exports.API_DIR = API_DIR;
const INTROSPECTION_ROUTE_METHOD = 'GET';
exports.INTROSPECTION_ROUTE_METHOD = INTROSPECTION_ROUTE_METHOD;
