// @modern-js/plugin-analyze@1.3.3/dist/js/node/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PAGES_DIR_NAME = exports.JS_EXTENSIONS = exports.INDEX_FILE_NAME = exports.HTML_PARTIALS_FOLDER = exports.HTML_PARTIALS_EXTENSIONS = exports.FILE_SYSTEM_ROUTES_LAYOUT = exports.FILE_SYSTEM_ROUTES_INDEX = exports.FILE_SYSTEM_ROUTES_IGNORED_REGEX = exports.FILE_SYSTEM_ROUTES_GLOBAL_LAYOUT = exports.FILE_SYSTEM_ROUTES_FILE_NAME = exports.FILE_SYSTEM_ROUTES_DYNAMIC_REGEXP = exports.FILE_SYSTEM_ROUTES_COMPONENTS_DIR = exports.ENTRY_POINT_FILE_NAME = exports.APP_FILE_NAME = void 0;
const JS_EXTENSIONS = ['.js', '.ts', '.jsx', '.tsx'];
exports.JS_EXTENSIONS = JS_EXTENSIONS;
const INDEX_FILE_NAME = 'index';
exports.INDEX_FILE_NAME = INDEX_FILE_NAME;
const APP_FILE_NAME = 'App';
exports.APP_FILE_NAME = APP_FILE_NAME;
const PAGES_DIR_NAME = 'pages';
exports.PAGES_DIR_NAME = PAGES_DIR_NAME;
const FILE_SYSTEM_ROUTES_FILE_NAME = 'routes.js';
exports.FILE_SYSTEM_ROUTES_FILE_NAME = FILE_SYSTEM_ROUTES_FILE_NAME;
const ENTRY_POINT_FILE_NAME = 'index.js';
exports.ENTRY_POINT_FILE_NAME = ENTRY_POINT_FILE_NAME;
const FILE_SYSTEM_ROUTES_DYNAMIC_REGEXP = /^\[(\S+)\]([*+?]?)$/;
exports.FILE_SYSTEM_ROUTES_DYNAMIC_REGEXP = FILE_SYSTEM_ROUTES_DYNAMIC_REGEXP;
const FILE_SYSTEM_ROUTES_LAYOUT = '_layout';
exports.FILE_SYSTEM_ROUTES_LAYOUT = FILE_SYSTEM_ROUTES_LAYOUT;
const FILE_SYSTEM_ROUTES_GLOBAL_LAYOUT = '_app';
exports.FILE_SYSTEM_ROUTES_GLOBAL_LAYOUT = FILE_SYSTEM_ROUTES_GLOBAL_LAYOUT;
const FILE_SYSTEM_ROUTES_INDEX = 'index';
exports.FILE_SYSTEM_ROUTES_INDEX = FILE_SYSTEM_ROUTES_INDEX;
const FILE_SYSTEM_ROUTES_IGNORED_REGEX = /\.(d|test|spec|e2e)\.(js|jsx|ts|tsx)$/;
exports.FILE_SYSTEM_ROUTES_IGNORED_REGEX = FILE_SYSTEM_ROUTES_IGNORED_REGEX;
const HTML_PARTIALS_FOLDER = 'html';
exports.HTML_PARTIALS_FOLDER = HTML_PARTIALS_FOLDER;
const HTML_PARTIALS_EXTENSIONS = ['.htm', '.html', '.ejs'];
exports.HTML_PARTIALS_EXTENSIONS = HTML_PARTIALS_EXTENSIONS;
const FILE_SYSTEM_ROUTES_COMPONENTS_DIR = 'internal_components';
exports.FILE_SYSTEM_ROUTES_COMPONENTS_DIR = FILE_SYSTEM_ROUTES_COMPONENTS_DIR;
