// @modern-js/runtime@1.2.3/dist/js/node/cli/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _path = _interopRequireDefault(require("path"));

var _utils = require("@modern-js/utils");

var _core = require("@modern-js/core");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const useInternalDirectory = () => (0, _core.useAppContext)().internalDirectory; // eslint-disable-next-line react-hooks/rules-of-hooks


(0, _core.usePlugins)([require.resolve('@modern-js/plugin-state/cli'), require.resolve('@modern-js/plugin-router/cli'), require.resolve('@modern-js/plugin-ssr/cli')]);

var _default = (0, _core.createPlugin)(() => {
  let runtimeExportsUtils = {};
  return {
    config() {
      // eslint-disable-next-line react-hooks/rules-of-hooks
      const dir = useInternalDirectory();
      runtimeExportsUtils = (0, _utils.createRuntimeExportsUtils)(dir, 'index');
      return {
        runtime: {},
        runtimeByEntries: {},
        source: {
          alias: {
            '@modern-js/runtime$': runtimeExportsUtils.getPath()
          }
        }
      };
    },

    validateSchema() {
      return _utils.PLUGIN_SCHEMAS['@modern-js/runtime'];
    },

    addRuntimeExports() {
      const runtimePackage = _path.default.resolve(__dirname, '../../../../');

      runtimeExportsUtils.addExport(`export * from '${runtimePackage}'`);
    },

    async beforeRestart() {
      (0, _utils.cleanRequireCache)([require.resolve('@modern-js/plugin-state/cli'), require.resolve('@modern-js/plugin-router/cli'), require.resolve('@modern-js/plugin-ssr/cli')]);
    }

  };
}, {
  name: '@modern-js/runtime',
  post: ['@modern-js/plugin-router', '@modern-js/plugin-ssr', '@modern-js/plugin-state', '@modern-js/plugin-design-token']
});

exports.default = _default;
