// @modern-js/utils@1.3.4/dist/js/modern/monorepo.js
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import fs from 'fs';
import path from 'path';
import glob from 'glob';
import yaml from 'yaml';
const PACKAGE_MAX_DEPTH = 5;
const WOKRSPACES_FILES = {
  YARN: 'package.json',
  PNPM: 'pnpm-workspaces.yaml',
  LERNA: 'lerna.json'
};
export const isLerna = root => fs.existsSync(path.join(root, WOKRSPACES_FILES.LERNA));
export const isYarnWorkspaces = root => {
  var _json$workspaces;

  const pkg = path.join(root, WOKRSPACES_FILES.YARN);

  if (!fs.existsSync(pkg)) {
    return false;
  }

  const json = JSON.parse(fs.readFileSync(pkg, 'utf8'));
  return Boolean((_json$workspaces = json.workspaces) === null || _json$workspaces === void 0 ? void 0 : _json$workspaces.packages);
};
export const isPnpmWorkspaces = root => fs.existsSync(path.join(root, WOKRSPACES_FILES.PNPM));
export const isMonorepo = root => isLerna(root) || isYarnWorkspaces(root) || isPnpmWorkspaces(root);
export const isModernjsMonorepo = root => {
  const json = JSON.parse(fs.readFileSync(path.join(root, 'package.json'), 'utf8'));

  const deps = _objectSpread(_objectSpread({}, json.dependencies || {}), json.devDependencies || {});

  return Boolean(deps['@modern-js/monorepo-tools']);
};
export const findMonorepoRoot = (appDirectory, maxDepth = PACKAGE_MAX_DEPTH) => {
  let inMonorepo = false;

  for (let depth = 0; depth < maxDepth; depth++) {
    if (isMonorepo(appDirectory)) {
      inMonorepo = true;
      break;
    } // eslint-disable-next-line no-param-reassign


    appDirectory = path.dirname(appDirectory);
  }

  return inMonorepo ? appDirectory : undefined;
};
export const getMonorepoPackages = root => {
  let packages = [];

  if (isYarnWorkspaces(root)) {
    const json = JSON.parse(fs.readFileSync(path.join(root, 'package.json'), 'utf8'));
    ({
      packages
    } = json.workspaces);
  } else if (isLerna(root)) {
    const json = JSON.parse(fs.readFileSync(path.resolve(root, 'lerna.json'), 'utf8'));
    ({
      packages
    } = json);
  } else {
    ({
      packages
    } = yaml.parse(fs.readFileSync(path.join(root, WOKRSPACES_FILES.PNPM), 'utf8')));
  }

  if (packages) {
    return packages.map(name => // The trailing / ensures only dirs are picked up
    glob.sync(path.join(root, `${name}/`), {
      ignore: ['**/node_modules/**']
    })).flat().filter(filepath => fs.existsSync(path.resolve(filepath, 'package.json'))).map(filepath => ({
      path: filepath,
      name: JSON.parse(fs.readFileSync(path.resolve(filepath, 'package.json'), 'utf8')).name
    }));
  }

  return [];
};
