// @modern-js/utils@1.3.4/dist/js/node/is/type.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isArray = isArray;
exports.isFunction = isFunction;
exports.isObject = isObject;
exports.isPlainObject = isPlainObject;
exports.isPromise = isPromise;
exports.isRegExp = isRegExp;
exports.isString = isString;
exports.isUndefined = isUndefined;

function isString(str) {
  return typeof str === 'string';
}

function isUndefined(obj) {
  return typeof obj === 'undefined';
}

function isArray(obj) {
  return Object.prototype.toString.call(obj) === '[object Array]';
} // eslint-disable-next-line @typescript-eslint/ban-types


function isFunction(func) {
  return typeof func === 'function';
} // eslint-disable-next-line @typescript-eslint/ban-types


function isObject(obj) {
  return typeof obj === 'object';
}

function isPlainObject(obj) {
  return obj && typeof obj === 'object' && Object.prototype.toString.call(obj) === '[object Object]';
}

function isPromise(obj) {
  /* eslint-disable promise/prefer-await-to-then */
  return Boolean(obj) && (typeof obj === 'object' || typeof obj === 'function') && typeof obj.then === 'function';
  /* eslint-enable promise/prefer-await-to-then */
}

function isRegExp(obj) {
  return Object.prototype.toString.call(obj) === '[object RegExp]';
}
