// @modern-js/plugin@1.2.1/dist/js/treeshaking/manager/async.js
function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

import _regeneratorRuntime from "@babel/runtime/regenerator";

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import { runWithContainer, createContainer } from 'farrow-pipeline';
import { generateRunner, hasOwnProperty, DEFAULT_OPTIONS } from "./sync";
import { useRunner } from "./runner"; // eslint-disable-next-line @typescript-eslint/no-invalid-void-type

var ASYNC_PLUGIN_SYMBOL = 'ASYNC_PLUGIN_SYMBOL';
export var createAsyncManager = function createAsyncManager(processes) {
  var index = 0;

  var createPlugin = function createPlugin(initializer) {
    var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    return _objectSpread(_objectSpread(_objectSpread({}, DEFAULT_OPTIONS), {}, {
      name: "No.".concat(index++, " plugin")
    }, options), {}, {
      ASYNC_PLUGIN_SYMBOL: ASYNC_PLUGIN_SYMBOL,
      initializer: initializer
    });
  };

  var isPlugin = function isPlugin(input) {
    return hasOwnProperty(input, ASYNC_PLUGIN_SYMBOL) && input[ASYNC_PLUGIN_SYMBOL] === ASYNC_PLUGIN_SYMBOL;
  };

  var registe = function registe(extraProcesses) {
    // eslint-disable-next-line no-param-reassign
    processes = _objectSpread(_objectSpread({}, extraProcesses), processes);
  };

  var clone = function clone() {
    var plugins = [];

    var usePlugin = function usePlugin() {
      for (var _len = arguments.length, input = new Array(_len), _key = 0; _key < _len; _key++) {
        input[_key] = arguments[_key];
      }

      for (var _i = 0, _input = input; _i < _input.length; _i++) {
        var plugin = _input[_i];

        if (isPlugin(plugin)) {
          if (!includeAsyncPlugin(plugins, plugin)) {
            plugins.push(_objectSpread(_objectSpread({}, plugin), {}, {
              index: plugins.length
            }));
          }
        } else {
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-expect-error
          console.warn("Unknown plugin: ".concat(plugin.name));
        }
      }

      return manager;
    };

    var clear = function clear() {
      plugins = [];
    };

    var currentContainer = createContainer();

    var init = /*#__PURE__*/function () {
      var _ref = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee(options) {
        var container, sortedPlugins, hooksList;
        return _regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                container = (options === null || options === void 0 ? void 0 : options.container) || currentContainer;
                sortedPlugins = sortAsyncPlugins(plugins);
                checkAsyncPlugins(sortedPlugins);
                _context.next = 5;
                return Promise.all(sortedPlugins.map(function (plugin) {
                  return runWithContainer(function () {
                    return plugin.initializer();
                  }, container);
                }));

              case 5:
                hooksList = _context.sent;
                return _context.abrupt("return", generateRunner(hooksList, container, processes));

              case 7:
              case "end":
                return _context.stop();
            }
          }
        }, _callee);
      }));

      return function init(_x) {
        return _ref.apply(this, arguments);
      };
    }();

    var run = function run(cb, options) {
      var container = (options === null || options === void 0 ? void 0 : options.container) || currentContainer;
      return runWithContainer(cb, container);
    };

    var manager = {
      createPlugin: createPlugin,
      isPlugin: isPlugin,
      usePlugin: usePlugin,
      init: init,
      run: run,
      clear: clear,
      clone: clone,
      registe: registe,
      useRunner: useRunner
    };
    return manager;
  };

  return clone();
};

var includeAsyncPlugin = function includeAsyncPlugin(plugins, input) {
  var _iterator = _createForOfIteratorHelper(plugins),
      _step;

  try {
    for (_iterator.s(); !(_step = _iterator.n()).done;) {
      var plugin = _step.value;

      if (plugin.name === input.name) {
        return true;
      }
    }
  } catch (err) {
    _iterator.e(err);
  } finally {
    _iterator.f();
  }

  return false;
};

var sortAsyncPlugins = function sortAsyncPlugins(input) {
  var plugins = input.slice();

  for (var i = 0; i < plugins.length; i++) {
    var plugin = plugins[i];

    var _iterator2 = _createForOfIteratorHelper(plugin.pre),
        _step2;

    try {
      for (_iterator2.s(); !(_step2 = _iterator2.n()).done;) {
        var pre = _step2.value;

        for (var j = i + 1; j < plugins.length; j++) {
          if (plugins[j].name === pre) {
            plugins = [].concat(_toConsumableArray(plugins.slice(0, i)), [plugins[j]], _toConsumableArray(plugins.slice(i, j)), _toConsumableArray(plugins.slice(j + 1, plugins.length)));
          }
        }
      }
    } catch (err) {
      _iterator2.e(err);
    } finally {
      _iterator2.f();
    }

    var _iterator3 = _createForOfIteratorHelper(plugin.post),
        _step3;

    try {
      for (_iterator3.s(); !(_step3 = _iterator3.n()).done;) {
        var post = _step3.value;

        for (var _j = 0; _j < i; _j++) {
          if (plugins[_j].name === post) {
            plugins = [].concat(_toConsumableArray(plugins.slice(0, _j)), _toConsumableArray(plugins.slice(_j + 1, i + 1)), [plugins[_j]], _toConsumableArray(plugins.slice(i + 1, plugins.length)));
          }
        }
      }
    } catch (err) {
      _iterator3.e(err);
    } finally {
      _iterator3.f();
    }
  }

  return plugins;
};

var checkAsyncPlugins = function checkAsyncPlugins(plugins) {
  var _iterator4 = _createForOfIteratorHelper(plugins),
      _step4;

  try {
    for (_iterator4.s(); !(_step4 = _iterator4.n()).done;) {
      var origin = _step4.value;

      var _iterator5 = _createForOfIteratorHelper(origin.rivals),
          _step5;

      try {
        for (_iterator5.s(); !(_step5 = _iterator5.n()).done;) {
          var rival = _step5.value;

          var _iterator7 = _createForOfIteratorHelper(plugins),
              _step7;

          try {
            for (_iterator7.s(); !(_step7 = _iterator7.n()).done;) {
              var plugin = _step7.value;

              if (rival === plugin.name) {
                throw new Error("".concat(origin.name, " has rival ").concat(plugin.name));
              }
            }
          } catch (err) {
            _iterator7.e(err);
          } finally {
            _iterator7.f();
          }
        }
      } catch (err) {
        _iterator5.e(err);
      } finally {
        _iterator5.f();
      }

      var _iterator6 = _createForOfIteratorHelper(origin.required),
          _step6;

      try {
        var _loop = function _loop() {
          var required = _step6.value;

          if (!plugins.some(function (plugin) {
            return plugin.name === required;
          })) {
            throw new Error("The plugin: ".concat(required, " is required when plugin: ").concat(origin.name, " is exist."));
          }
        };

        for (_iterator6.s(); !(_step6 = _iterator6.n()).done;) {
          _loop();
        }
      } catch (err) {
        _iterator6.e(err);
      } finally {
        _iterator6.f();
      }
    }
  } catch (err) {
    _iterator4.e(err);
  } finally {
    _iterator4.f();
  }
};
