// @modern-js/webpack@1.4.1/dist/js/node/utils/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TS_REGEX = exports.SVG_REGEX = exports.JS_RESOLVE_EXTENSIONS = exports.JS_REGEX = exports.ICON_EXTENSIONS = exports.GLOBAL_CSS_REGEX = exports.CSS_REGEX = exports.CSS_MODULE_REGEX = exports.CACHE_DIRECTORY = exports.ASSETS_REGEX = void 0;
const CSS_REGEX = /\.css$/;
exports.CSS_REGEX = CSS_REGEX;
const CSS_MODULE_REGEX = /\.module\.css$/;
exports.CSS_MODULE_REGEX = CSS_MODULE_REGEX;
const GLOBAL_CSS_REGEX = /\.global\.css$/;
exports.GLOBAL_CSS_REGEX = GLOBAL_CSS_REGEX;
const JS_REGEX = /\.(js|mjs|jsx)$/;
exports.JS_REGEX = JS_REGEX;
const TS_REGEX = /\.tsx?$/;
exports.TS_REGEX = TS_REGEX;
const ASSETS_REGEX = /\.(woff|woff2|eot|ttf|otf|ttc|gif|png|jpe?g|webp|bmp|ico)$/i;
exports.ASSETS_REGEX = ASSETS_REGEX;
const ICON_EXTENSIONS = ['png', 'jpg', 'jpeg', 'svg', 'ico'];
exports.ICON_EXTENSIONS = ICON_EXTENSIONS;
const SVG_REGEX = /\.svg$/;
exports.SVG_REGEX = SVG_REGEX;
const JS_RESOLVE_EXTENSIONS = ['web.mjs', 'mjs', 'web.js', 'js', 'web.ts', 'ts', 'web.tsx', 'tsx', 'json', 'web.jsx', 'jsx'].map(t => `.${t}`);
exports.JS_RESOLVE_EXTENSIONS = JS_RESOLVE_EXTENSIONS;
const CACHE_DIRECTORY = './node_modules/.cache';
exports.CACHE_DIRECTORY = CACHE_DIRECTORY;
