// @modern-js/codesmith-api-npm@1.0.7/dist/js/node/utils/env.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.canUseNpm = canUseNpm;
exports.canUseNvm = canUseNvm;
exports.canUsePnpm = canUsePnpm;
exports.canUseYarn = canUseYarn;

var _execa = _interopRequireDefault(require("execa"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

async function canUseNvm() {
  try {
    await (0, _execa.default)('nvm --version', {
      env: process.env,
      shell: true
    });
    return true;
  } catch (e) {
    return false;
  }
}

async function canUseNpm() {
  try {
    await (0, _execa.default)('npm', ['--version'], {
      env: process.env
    });
    return true;
  } catch (e) {
    return false;
  }
}

async function canUseYarn() {
  try {
    await (0, _execa.default)('yarn', ['--version'], {
      env: process.env
    });
    return true;
  } catch (e) {
    return false;
  }
}

async function canUsePnpm() {
  try {
    await (0, _execa.default)('pnpm', ['--version'], {
      env: process.env
    });
    return true;
  } catch (e) {
    return false;
  }
}
