// @modern-js/codesmith-api-npm@1.0.7/dist/js/node/utils/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "canUseNpm", {
  enumerable: true,
  get: function () {
    return _env.canUseNpm;
  }
});
Object.defineProperty(exports, "canUseNvm", {
  enumerable: true,
  get: function () {
    return _env.canUseNvm;
  }
});
Object.defineProperty(exports, "canUsePnpm", {
  enumerable: true,
  get: function () {
    return _env.canUsePnpm;
  }
});
Object.defineProperty(exports, "canUseYarn", {
  enumerable: true,
  get: function () {
    return _env.canUseYarn;
  }
});
Object.defineProperty(exports, "npmInstall", {
  enumerable: true,
  get: function () {
    return _install.npmInstall;
  }
});
Object.defineProperty(exports, "pnpmInstall", {
  enumerable: true,
  get: function () {
    return _install.pnpmInstall;
  }
});
Object.defineProperty(exports, "yarnInstall", {
  enumerable: true,
  get: function () {
    return _install.yarnInstall;
  }
});

var _env = require("./env");

var _install = require("./install");
